runtime/nsd_gtls.c

```diff
@@ -576,14 +577,31 @@ gtlsRecordRecv(nsd_gtls_t *pThis)
 					(NSD_GTLS_MAX_RCVBUF+lenRcvd));
 				pThis->lenRcvBuf = NSD_GTLS_MAX_RCVBUF+lenRcvd;
 			} else {
-				goto sslerr;
+				if (lenRcvd == GNUTLS_E_AGAIN || lenRcvd == GNUTLS_E_INTERRUPTED) {
+					goto sslerragain;	/* Go to ERR AGAIN handling */
+				} else {
+					/* Do all other error handling */
+					int gnuRet = lenRcvd;
+					ABORTgnutls;
+				}
 			}
 		}
 	} else if(lenRcvd == GNUTLS_E_AGAIN || lenRcvd == GNUTLS_E_INTERRUPTED) {
-sslerr:
-		pThis->rtryCall = gtlsRtry_recv;
-		dbgprintf("GnuTLS receive requires a retry (this most probably is OK and no error condition)\n");
-		ABORT_FINALIZE(RS_RET_RETRY);
+sslerragain:
+		/* Check if the underlaying file descriptor needs to read or write data!*/
+		wantsWriteData = gnutls_record_get_direction(pThis->sess);
+		if (wantsWriteData == 0) {
```

Comment:
In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

Reply:
@rgerhards what do you think? In order to maintain existing 

> In our case, the problem is the thread keeps doing 'recvfrom()' endlessly. Could we abort the connection after some retries or some time?

Could you post a gdb stack trace?
Or even better could you enable debug output, if we could catch that error with debug log enabled that would be awesome.